designs/sam-debugging/local-sam-debugging.md

```diff
@@ -0,0 +1,308 @@
+
+These CodeLenses do not support API Gateway style invokes.
+
+Some users may find CodeLenses within code files distracting, particularly if they are using the Toolkit for features not related to local debugging. Toolkit settings can be used to enable and disable CodeLenses. CodeLenses only appear for languages/runtimes that the Toolkit has implemented Debug support for.
+
+### Graphical User Interface
+
+SAM Template Resources invoked by API Gateway can be locally tested with a REST Workbench.
+
+The REST Workbench is accessed with the command "Open Local REST Workbench for SAM Applications", opens a webview document.
+
+In the REST Workbench, users select a SAM Application Template, build and launch the application (with or without debug mode), and proceed to make a series of REST requests against the running application.
+
+![Workbench Start-up](workbench-start.png)
+
+Users start by selecting the SAM Application they would like to work with. A dropdown lists all SAM Template files detected within the workspace. Next, they launch the application (Run or Debug), which builds and launches the local web server host. At this point, the SAM Application can no longer be changed (users can open a new workbench in order to operate against a different SAM Application).
+
+![Workbench Request Iteration](workbench-request.png)
+
+The running instance of the application is used for all REST queries made by the user. Users define REST parameters (endpoint, request type (GET, POST, etc), a query string, and an optional request body), and press Submit. If the application is run in Debug mode, the Toolkit will attach a debugger after sending the request. The following information is sent to the Toolkit's Output Channel with each REST request:
+
+-   request
+-   response
+-   response code
+-   SAM CLI output
+
+After making code changes, users can click Restart to rebuild the application and start it again.
+
+Closing a REST Workbench tab closes the invoked SAM Application if one is running.
+
+## Appendix
+
+### Differences from v1.0.0 of AWS Toolkit
+
+The debug capabilities initially released in the Toolkit were not well rounded. CodeLenses provided the only means of local debugging, and without a way to directly debug SAM Template Resources these CodeLenses tried to compensate, leading to scenarios with undefined behaviors. Many of the issues are referenced from https://github.com/aws/aws-toolkit-vscode/issues/758
+
+Here is an outline of the differences between this design and version 1.0.0 of the AWS Toolkit:
+
+-   Changed functionality
+    -   CodeLenses on code files invoke the function in isolation, and no longer attempt to associate the function with a SAM Template
```

Comment:
Could this lead to confusing behaviour where previously a customer was expecting their other resources to be in play?

Reply:
It could. Users will appreciate having access to F5 and the Debug pane to invoke now, instead of hunting for their file/function as the execution point. Current functionality does not support 'naked' lambda execution.

This needs to be mitigated through changelog entries, user docs, and inline where appropriate (tooltips for example).